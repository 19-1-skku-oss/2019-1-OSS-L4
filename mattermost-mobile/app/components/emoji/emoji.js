// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import React from 'react';
import PropTypes from 'prop-types';
import {
    Image,
    Platform,
    StyleSheet,
    Text,
} from 'react-native';

import CustomPropTypes from 'app/constants/custom_prop_types';
import ImageCacheManager from 'app/utils/image_cache_manager';

export default class Emoji extends React.PureComponent {
    static propTypes = {

        /*
         * Emoji text name.
         */
        emojiName: PropTypes.string.isRequired,

        /*
         * Image URL for the emoji.
         */
        imageUrl: PropTypes.string.isRequired,

        /*
         * Set if this is a custom emoji.
         */
        isCustomEmoji: PropTypes.bool.isRequired,

        /*
         * Set to render only the text and no image.
         */
        displayTextOnly: PropTypes.bool,
        literal: PropTypes.string,
        size: PropTypes.number,
        textStyle: CustomPropTypes.Style,
    };

    static defaultProps = {
        customEmojis: new Map(),
        literal: '',
        imageUrl: '',
        isCustomEmoji: false,
    };

    constructor(props) {
        super(props);

        this.state = {
            imageUrl: null,
        };
    }

    componentWillMount() {
        const {displayTextOnly, emojiName, imageUrl} = this.props;
        this.mounted = true;
        if (!displayTextOnly && imageUrl) {
            ImageCacheManager.cache(`emoji-${emojiName}`, imageUrl, this.setImageUrl);
        }
    }

    componentWillReceiveProps(nextProps) {
        const {displayTextOnly, emojiName, imageUrl} = nextProps;
        if (emojiName !== this.props.emojiName && this.mounted) {
            this.setState({
                imageUrl: null,
            });
        }

        if (!displayTextOnly && imageUrl &&
                imageUrl !== this.props.imageUrl) {
            ImageCacheManager.cache(`emoji-${emojiName}`, imageUrl, this.setImageUrl);
        }
    }

    componentWillUnmount() {
        this.mounted = false;
    }

    setImageUrl = (imageUrl) => {
        if (this.mounted) {
            this.setState({
                imageUrl,
            });
        }
    };

    render() {
        const {
            literal,
            textStyle,
            displayTextOnly,
        } = this.props;
        const {imageUrl} = this.state;

        let size = this.props.size;
        let fontSize = size;
        if (!size && textStyle) {
            const flatten = StyleSheet.flatten(textStyle);
            fontSize = flatten.fontSize;
            size = fontSize;
        }

        if (displayTextOnly) {
            return <Text style={textStyle}>{literal}</Text>;
        }

        const width = size;
        const height = size;

        // Android can't change the size of an image after its first render, so
        // force a new image to be rendered when the size changes
        const key = Platform.OS === 'android' ? (height + '-' + width) : null;

        if (!imageUrl) {
            return (
                <Image
                    key={key}
                    style={{width, height}}
                />
            );
        }

        return (
            <Image
                key={key}
                style={{width, height}}
                source={{uri: imageUrl}}
                onError={this.onError}
            />
        );
    }
}
