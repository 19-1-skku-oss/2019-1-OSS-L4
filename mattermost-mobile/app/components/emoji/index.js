// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import {connect} from 'react-redux';

import {getCustomEmojisByName} from 'mattermost-redux/selectors/entities/emojis';
import {getCurrentUserId} from 'mattermost-redux/selectors/entities/users';
import {getConfig} from 'mattermost-redux/selectors/entities/general';
import {Client4} from 'mattermost-redux/client';
import {isMinimumServerVersion} from 'mattermost-redux/utils/helpers';

import {EmojiIndicesByAlias, Emojis} from 'app/utils/emojis';

import Emoji from './emoji';

function mapStateToProps(state, ownProps) {
    const config = getConfig(state);
    const emojiName = ownProps.emojiName;
    const customEmojis = getCustomEmojisByName(state);

    let imageUrl = '';
    let isCustomEmoji = false;
    let displayTextOnly = false;
    if (EmojiIndicesByAlias.has(emojiName)) {
        const emoji = Emojis[EmojiIndicesByAlias.get(emojiName)];
        imageUrl = Client4.getSystemEmojiImageUrl(emoji.filename);
    } else if (customEmojis.has(emojiName)) {
        const emoji = customEmojis.get(emojiName);
        imageUrl = Client4.getCustomEmojiImageUrl(emoji.id);
        isCustomEmoji = true;
    } else {
        displayTextOnly = state.entities.emojis.nonExistentEmoji.has(emojiName) ||
            config.EnableCustomEmoji !== 'true' ||
            config.ExperimentalEnablePostMetadata === 'true' ||
            getCurrentUserId(state) === '' ||
            !isMinimumServerVersion(Client4.getServerVersion(), 4, 7);
    }

    return {
        imageUrl,
        isCustomEmoji,
        displayTextOnly,
    };
}

export default connect(mapStateToProps)(Emoji);
