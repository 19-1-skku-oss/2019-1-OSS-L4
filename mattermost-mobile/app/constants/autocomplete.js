// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

export const AT_MENTION_REGEX = /\B(@([^@\r\n\s]*))$/i;

export const AT_MENTION_SEARCH_REGEX = /\bfrom:\s*(\S*)$/i;

export const CHANNEL_MENTION_REGEX = /\B(~([^~\r\n]*))$/i;

export const CHANNEL_MENTION_SEARCH_REGEX = /\b(?:in|channel):\s*(\S*)$/i;

export const DATE_MENTION_SEARCH_REGEX = /\b(?:on|before|after):\s*(\S*)$/i;

export const ALL_SEARCH_FLAGS_REGEX = /\b\w+:/g;