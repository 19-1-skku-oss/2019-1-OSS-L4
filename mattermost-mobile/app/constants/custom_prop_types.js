// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import PropTypes from 'prop-types';

export const Children = PropTypes.oneOfType([PropTypes.node, PropTypes.arrayOf([PropTypes.node])]);

export const Style = PropTypes.oneOfType([
    PropTypes.object, // inline style
    PropTypes.number, // style sheet entry
    PropTypes.array,
]);

export default {
    Children,
    Style,
};
