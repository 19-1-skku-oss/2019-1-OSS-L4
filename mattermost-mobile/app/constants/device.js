// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import DeviceInfo from 'react-native-device-info';
import RNFetchBlobFS from 'rn-fetch-blob/fs';
import keyMirror from 'mattermost-redux/utils/key_mirror';

const deviceTypes = keyMirror({
    CONNECTION_CHANGED: null,
    DEVICE_DIMENSIONS_CHANGED: null,
    DEVICE_TYPE_CHANGED: null,
    DEVICE_ORIENTATION_CHANGED: null,
    STATUSBAR_HEIGHT_CHANGED: null,
});

export default {
    ...deviceTypes,
    DOCUMENTS_PATH: `${RNFetchBlobFS.dirs.CacheDir}/Documents`,
    IMAGES_PATH: `${RNFetchBlobFS.dirs.CacheDir}/Images`,
    IS_IPHONE_X: DeviceInfo.getModel().includes('iPhone X'),
    IS_TABLET: DeviceInfo.isTablet(),
    VIDEOS_PATH: `${RNFetchBlobFS.dirs.CacheDir}/Videos`,
};
