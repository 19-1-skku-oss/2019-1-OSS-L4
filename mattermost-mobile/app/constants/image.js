// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

export const IMAGE_MAX_HEIGHT = 350;
export const IMAGE_MIN_DIMENSION = 50;
