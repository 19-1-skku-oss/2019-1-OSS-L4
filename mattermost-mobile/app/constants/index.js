// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import DeepLinkTypes from './deep_linking';
import DeviceTypes from './device';
import ListTypes from './list';
import NavigationTypes from './navigation';
import PermissionTypes from './permissions';
import ViewTypes, {UpgradeTypes} from './view';

export {
    DeepLinkTypes,
    DeviceTypes,
    ListTypes,
    NavigationTypes,
    PermissionTypes,
    UpgradeTypes,
    ViewTypes,
};
