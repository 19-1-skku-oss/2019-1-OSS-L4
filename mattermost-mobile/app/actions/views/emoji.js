// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import {addReaction as serviceAddReaction} from 'mattermost-redux/actions/posts';
import {getPostIdsInCurrentChannel, makeGetPostIdsForThread} from 'mattermost-redux/selectors/entities/posts';

import {ViewTypes} from 'app/constants';

const getPostIdsForThread = makeGetPostIdsForThread();

export function addReaction(postId, emoji) {
    return (dispatch) => {
        dispatch(serviceAddReaction(postId, emoji));
        dispatch(addRecentEmoji(emoji));
    };
}

export function addReactionToLatestPost(emoji, rootId) {
    return async (dispatch, getState) => {
        const state = getState();
        const postIds = rootId ? getPostIdsForThread(state, rootId) : getPostIdsInCurrentChannel(state);
        const lastPostId = postIds[0];

        dispatch(serviceAddReaction(lastPostId, emoji));
        dispatch(addRecentEmoji(emoji));
    };
}

export function addRecentEmoji(emoji) {
    return {
        type: ViewTypes.ADD_RECENT_EMOJI,
        emoji,
    };
}

export function incrementEmojiPickerPage() {
    return async (dispatch) => {
        dispatch({
            type: ViewTypes.INCREMENT_EMOJI_PICKER_PAGE,
        });

        return {data: true};
    };
}
