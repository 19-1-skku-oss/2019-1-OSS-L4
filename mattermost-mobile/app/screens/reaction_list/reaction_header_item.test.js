// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.
import React from 'react';
import {shallow} from 'enzyme';
import {TouchableOpacity} from 'react-native';

import Preferences from 'mattermost-redux/constants/preferences';

import ReactionHeaderItem from './reaction_header_item';

import {ALL_EMOJIS} from 'app/constants/emoji';

describe('ReactionHeaderItem', () => {
    const baseProps = {
        count: 3,
        emojiName: 'smile',
        highlight: false,
        onPress: jest.fn(),
        theme: Preferences.THEMES.default,
    };

    test('should match snapshot', () => {
        const wrapper = shallow(
            <ReactionHeaderItem {...baseProps}/>
        );

        expect(wrapper.getElement()).toMatchSnapshot();
        expect(wrapper.find(TouchableOpacity).exists()).toEqual(true);
    });

    test('should match snapshot, renderContent', () => {
        const wrapper = shallow(
            <ReactionHeaderItem {...baseProps}/>
        );

        expect(wrapper.instance().renderContent()).toMatchSnapshot();

        wrapper.setProps({emojiName: ALL_EMOJIS});
        expect(wrapper.instance().renderContent()).toMatchSnapshot();
    });

    test('should call props.onPress on handleOnPress', () => {
        const onPress = jest.fn();
        const wrapper = shallow(
            <ReactionHeaderItem
                {...baseProps}
                onPress={onPress}
            />
        );

        wrapper.instance().handleOnPress();
        expect(onPress).toHaveBeenCalledTimes(1);
        expect(onPress).toHaveBeenCalledWith(baseProps.emojiName, baseProps.highlight);
    });
});
