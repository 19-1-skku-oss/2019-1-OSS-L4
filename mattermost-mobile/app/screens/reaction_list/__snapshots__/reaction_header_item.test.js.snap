// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ReactionHeaderItem should match snapshot 1`] = `
<TouchableOpacity
  activeOpacity={0.2}
  onPress={[Function]}
  style={
    Array [
      Object {
        "alignItems": "center",
        "flexDirection": "row",
        "height": 35,
        "marginBottom": 5,
        "marginRight": 6,
        "marginTop": 3,
        "paddingHorizontal": 6,
        "paddingVertical": 2,
      },
      Object {
        "borderBottomWidth": 2,
        "borderColor": "transparent",
      },
      false,
    ]
  }
>
  <React.Fragment>
    <Connect(Emoji)
      emojiName="smile"
      padding={5}
      size={16}
    />
    <Text
      style={
        Object {
          "color": "#166de0",
          "fontSize": 16,
          "marginLeft": 4,
        }
      }
    >
      3
    </Text>
  </React.Fragment>
</TouchableOpacity>
`;

exports[`ReactionHeaderItem should match snapshot, renderContent 1`] = `
<React.Fragment>
  <Connect(Emoji)
    emojiName="smile"
    padding={5}
    size={16}
  />
  <Text
    style={
      Object {
        "color": "#166de0",
        "fontSize": 16,
        "marginLeft": 4,
      }
    }
  >
    3
  </Text>
</React.Fragment>
`;

exports[`ReactionHeaderItem should match snapshot, renderContent 2`] = `
<React.Fragment>
  <FormattedText
    defaultMessage="All"
    id="mobile.reaction_header.all_emojis"
    style={
      Object {
        "color": "#166de0",
        "fontSize": 16,
        "marginLeft": 4,
      }
    }
  />
  <Text
    style={
      Object {
        "color": "#166de0",
        "fontSize": 16,
        "marginLeft": 4,
      }
    }
  >
    3
  </Text>
</React.Fragment>
`;
